jep/0000/README.adoc

```diff
@@ -0,0 +1,220 @@
+and plugin maintainer are able to detect such compatibility breaks, and provide some diagnostic and
+resolution hints.
+
+== Specification
+
+Configuration-as-Code tooling is designed to fill the gap between the raw mechanism defined by JEP-201
+to configure a specific jenkins-core + plugins combination, and the need to keep this configuration
+stable during core or plugin upgrades, or at least detect breaking changes and provide hints to manage
+upgreade.
+
+This includes:
+
+- tooling to validate configuration, sort of a "dry-run" mode, with adequate diagnostic when some
+configuration mismatch is detected.
+- tooling to compare inferred model between versions, to be used by plugin developers to be aware
+changes they made during development might have impact on Configuration-as-Code.
+- Document recommendations for plugin developers on ways to avoid breaking changes
+- Metadata definition for plugin developers to document breaking changes and how to fix them. Based
+on this, when yaml validation do detect a configuration mismatch it can offer migration hints.
+
+
+== Motivation
+
+Configuration-as-Code doesn't define a data model to setup jenkins, it's based on inferring a model
+from live jenkins instance from it's Java API, especially UI data-binding mechanisms. Based on this,
+changes made to core or a plugin _might_ change this model and a configuration file _could_ not be
+valid after upgrade.
+
+There's not so much cases where such compatibility breaks could happen, but there's chances this would.
+Compared to other tools, like `docker-compose`, we can't provide a `version` attribute and accept
+distinct models, as we only know about the live, inferred one.
+
+Configuration-as-Code tooling is defined to workaround this limitation by providing a toolset to
+discover such breaking changes early, before the actual production configuration is updated.
+It will help detect breaking changes during jenkins and plugins developement, offer an option for
+plugin developers to document required breaking changes, and validate a configuration for a specific
+jenkins instance before it get actually applied.
+
+== Reasoning
+
+Without an explicit versioned model per jenkins-core and plugin versions Configuration-as-Code only
+can rely offer some minimal backward compatibility.
+
+
+=== Relax Configuration-asCode for backward compatibility ?
+
+There's few circumstances where we have options
+to workaround a technical change in Jenkins or plugin codebase as the technical bits are still available,
+just hidden by the default "model inference" mechanism.
+
+A typical sample for this has been debated on
+https://github.com/jenkinsci/configuration-as-code-plugin/issues/233[issue 233]: `JNLPLauncher` DataBound
+constructor has been updated in jenkins codebase to introduce new argument `workDirSettings`. This one
+is declared in API as `@Nonnull`, which means Configuration-as-Code will infer a required attribute. As
```

Comment:
Another evaluation would be that this change to `JNLPLauncher` was wrong and should be fixed: if there was a reasonable default for `workDirSettings`, then it should have been in a `@DataBoundSetter` and thus optional. In fact there _is_ such a [reasonable default](https://github.com/jenkinsci/jenkins/blob/d4737b2b60438ce0e3cb7a8a6f26675f7c880914/core/src/main/java/jenkins/slaves/RemotingWorkDirSettings.java#L55) as seen also [here](https://github.com/jenkinsci/jenkins/blob/d4737b2b60438ce0e3cb7a8a6f26675f7c880914/core/src/main/resources/jenkins/slaves/RemotingWorkDirSettings/config.jelly#L28-L37). So this was just a bug in core that should be fixed, not something to work around. CC @oleg-nenashev